transaction/core/src/ring_signature/rct_bulletproofs.rs

```diff
@@ -308,35 +446,110 @@ fn sign_with_balance_check<CSPRNG: RngCore + CryptoRng>(
     let pseudo_output_values_and_blindings: Vec<(u64, Scalar)> = input_secrets
         .iter()
         .zip(pseudo_output_blindings.iter())
-        .map(|((_, value, _), blinding)| (*value, *blinding))
+        .map(|(secret, blinding)| (secret.value, *blinding))
         .collect();
 
-    // Compute the pedersen generators for this token_id
-    let generator = generators(token_id);
+    // Create a pedersen generator cache
+    let mut generator_cache = GeneratorCache::default();
 
-    let (range_proof, commitments) = {
+    // Range proof is present when mixed transactions are not supported, the set of
+    // range proofs is present when they are.
+    let (range_proof, range_proofs) = if !block_version.mixed_transactions_are_supported() {
         // The implicit fee output is omitted from the range proof because it is known.
+        let generator = generator_cache.get(fee_token_id);
 
         let (values, blindings): (Vec<_>, Vec<_>) = pseudo_output_values_and_blindings
             .iter()
-            .chain(output_values_and_blindings.iter())
-            .map(|(value, blinding)| (*value, *blinding))
+            .cloned()
+            .chain(
+                output_secrets
+                    .iter()
+                    .map(|secret| (secret.value, secret.blinding)),
+            )
             .unzip();
-        generate_range_proofs(&values, &blindings, &generator, rng)
-            .map_err(|_e| Error::RangeProofError)?
+        let (range_proof, _commitments) =
```

Comment:
I'm wondering about discarding the commitments here and want to check my understanding. You are simply regenerating them below? Is there a cost to doing that?

Reply:
yeah i'm regenerating them below

the thing is, in the old way, there was only one range proof, and it proves each of the commitments in order of pseudo_outputs, followed by outputs.

in the new world, there have to be multiple range proofs, one for each pedersen generator.
so I have to like, sort them and pick out the ones that match each token id.

if we try to take the commitments returned by `generate_range_proofs`, we have to "unsort" them back into the right order to be useful later, because the order of the pseudo output commtiments has to match the order of the TxIn in the prefix, which is sorted by tx public key.

i felt like it was simpler to just regenerate them, at least when trying to get the code to pass tests.
we can try to go back and do the "unsorting" thing, it will probably save some elliptic curve operations if we don't regenerate the commitments.

Reply:
If you don't mind, I would like to try to develop this unsorting in a subsequent commit, because it's going to make the signed contingent input commit more complicated

Reply:
i think the unsorting thing only impacts signing a transaction, so any slowdown is only client side and won't affect the performance of the network